core/drivers/crypto/se050/core/rsa.c

```diff
@@ -0,0 +1,705 @@
+					     &key_bin, key_bin.n_len * 8);
+	free(key_bin.n);
+	free(key_bin.e);
+
+	if (st != kStatus_SSS_Success) {
+		sss_se05x_key_store_erase_key(se050_kstore, k_object);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result se050_inject_keypair(sss_se05x_object_t *k_object,
+				       struct rsa_keypair *key)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	struct se050_rsa_keypair key_bin = { 0 };
+	uint32_t key_id = 0;
+	uint32_t oid = 0;
+
+	st = sss_se05x_key_object_init(k_object, se050_kstore);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	key_id = se050_rsa_keypair_from_nvm(key);
+	if (key_id) {
+		st = sss_se05x_key_object_get_handle(k_object, key_id);
+		if (st != kStatus_SSS_Success) {
+			EMSG("RSA");
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+		return TEE_SUCCESS;
+	}
+
+	st = se050_get_oid(kKeyObject_Mode_Transient, &oid);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_GENERIC;
+
+	st = sss_se05x_key_object_allocate_handle(k_object, oid,
+						  kSSS_KeyPart_Pair,
+						  kSSS_CipherType_RSA, 0,
+						  kKeyObject_Mode_Transient);
+	if (st != kStatus_SSS_Success)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	set_binary_data(key->e, &key_bin.e, &key_bin.e_len);
+	set_binary_data(key->d, &key_bin.d, &key_bin.d_len);
+	set_binary_data(key->n, &key_bin.n, &key_bin.n_len);
+	set_binary_data(key->p, &key_bin.p, &key_bin.p_len);
+	set_binary_data(key->q, &key_bin.q, &key_bin.q_len);
+	set_binary_data(key->qp, &key_bin.qp, &key_bin.qp_len);
+	set_binary_data(key->dp, &key_bin.dp, &key_bin.dp_len);
+	set_binary_data(key->dq, &key_bin.dq, &key_bin.dq_len);
+	st = se050_key_store_set_rsa_key_bin(se050_kstore, k_object,
+					     &key_bin, NULL,
+					     crypto_bignum_num_bytes(key->n)
+					     * 8);
+	free(key_bin.e);
+	free(key_bin.d);
+	free(key_bin.n);
+	free(key_bin.p);
+	free(key_bin.q);
+	free(key_bin.qp);
+	free(key_bin.dp);
+	free(key_bin.dq);
+
+	if (st != kStatus_SSS_Success) {
+		sss_se05x_key_store_erase_key(se050_kstore, k_object);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result decrypt_es(uint32_t algo, struct rsa_keypair *key,
+			     const uint8_t *src, size_t src_len,
+			     uint8_t *dst, size_t *dst_len)
+{
+	sss_status_t st = kStatus_SSS_Fail;
+	sss_se05x_asymmetric_t ctx = { 0 };
+	sss_se05x_object_t kobject = { 0 };
+	TEE_Result res = TEE_SUCCESS;
+
+	res = se050_inject_keypair(&kobject, key);
+	if (res != TEE_SUCCESS)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	st = sss_se05x_asymmetric_context_init(&ctx, se050_session, &kobject,
+					       tee2se050(algo),
+					       kMode_SSS_Decrypt);
+	if (st != kStatus_SSS_Success) {
+		if (!se050_rsa_keypair_from_nvm(key))
+			sss_se05x_key_store_erase_key(se050_kstore, &kobject);
+
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	st = sss_se05x_asymmetric_decrypt(&ctx, src, src_len, dst, dst_len);
+	if (st != kStatus_SSS_Success)
+		res = TEE_ERROR_BAD_PARAMETERS;
```

Comment:
Why BAD_PARAMETERS?

Reply:
how would you determine all the other possible errors (ie, i2c bus down)?